ostd/src/arch/x86/kernel/apic/mod.rs

```diff
@@ -1,22 +1,82 @@
 // SPDX-License-Identifier: MPL-2.0
 
-#![allow(dead_code)]
-
-use alloc::sync::Arc;
+use alloc::boxed::Box;
+use core::cell::UnsafeCell;
 
 use bit_field::BitField;
-use log::info;
 use spin::Once;
 
-use crate::sync::SpinLock;
+use crate::cpu_local;
 
 pub mod ioapic;
 pub mod x2apic;
 pub mod xapic;
 
-pub static APIC_INSTANCE: Once<Arc<SpinLock<dyn Apic + 'static>>> = Once::new();
+cpu_local! {
+    static APIC_INSTANCE: Once<Box<UnsafeCell<dyn Apic + 'static>>> = Once::new();
+}
+
 static APIC_TYPE: Once<ApicType> = Once::new();
 
+/// Do something over the APIC instance for the current CPU.
+///
+/// You should provide a closure operating on the given mutable borrow of the
+/// local APIC instance. During the execution of the closure, the interrupts
+/// are guarenteed to be disabled.
+///
+/// Example:
+/// ```rust
+/// use ostd::arch::x86::kernel::apic;
+///
+/// let ticks = apic::borrow(|apic| {
+///     let ticks = apic.timer_current_count();
+///     apic.set_timer_init_count(0);
+///     ticks
+/// });
+/// ```
+pub fn borrow<'a, R>(f: impl FnOnce(&'a mut (dyn Apic + 'static)) -> R) -> R {
+    let apic_guard = APIC_INSTANCE.borrow_irq_disabled();
+
+    // If it is not initialzed, lazily initialize it.
+    if !apic_guard.is_completed() {
+        apic_guard.call_once(|| match APIC_TYPE.get().unwrap() {
+            ApicType::XApic => {
+                let mut xapic = xapic::XApic::new().unwrap();
+                xapic.enable();
+                let version = xapic.version();
+                log::info!(
+                    "xAPIC ID:{:x}, Version:{:x}, Max LVT:{:x}",
+                    xapic.id(),
+                    version & 0xff,
+                    (version >> 16) & 0xff
+                );
+                Box::new(UnsafeCell::new(xapic))
+            }
+            ApicType::X2Apic => {
+                let mut x2apic = x2apic::X2Apic::new().unwrap();
+                x2apic.enable();
+                let version = x2apic.version();
+                log::info!(
+                    "x2APIC ID:{:x}, Version:{:x}, Max LVT:{:x}",
+                    x2apic.id(),
+                    version & 0xff,
+                    (version >> 16) & 0xff
+                );
+                Box::new(UnsafeCell::new(x2apic))
+            }
+        });
+    }
+
+    let apic = apic_guard.get().unwrap();
+
+    // SAFETY: the APIC instance is initialized and it is CPU local.
+    // We can safely get a mutable reference to the APIC instance when
+    // interrupts are disabled.
+    let apic = unsafe { &mut *apic.get() };
```

Comment:
I just came here to learn how we start APs. I don't have any more comments on this PR, except that there seems to be a small one.

I think this API is unsound because if we write `apic::borrow(|apic| some_func())` and `some_func` also calls `apic::borrow`, it will create multiple mutable references to the APIC instance.

I think a simple and quick fix is just to replace `UnsafeCell` with `Cell`. Does it work?

Reply:
Oh, you are right.

Reply:
Well, it seems to be more tricky than a `Cell` can handle. Since APIC users may modify the state of the APIC and then read a value out of it, we must allow the closure to return a generic value.

So I implemented a borrow checker in this method, and require the return value `R` to implement `Copy`. Then the closure cannot not leak a mutable reference, and the mutable borrow would be exclusive within this function.

I see it sound now. Do you feel it working?

Reply:
> Well, it seems to be more tricky than a Cell can handle. Since APIC users may modify the state of the APIC and then read a value out of it, we must allow the closure to return a generic value.

It isn't possible to reference the lifetime `'a` in `R`, right? Could you please give an example to show how this is unsound? I'm _very_ surprised if it is possible.

I tried [this example in the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=543d0a99b8ddbaea903b3190ed6ee7dc). It fails to compile because "lifetime may not live long enough":
```rust
fn foo<R>(_bar: impl FnOnce(&mut usize) -> R) -> R {
    todo!();
}

fn main() {
    foo(|the_ref| the_ref);
}
```

> So I implemented a borrow checker in this method,

Sorry, but I don't understand why you need to implement the borrow checker manually. Could you please explain a bit more?

Reply:
> I don't understand why you need to implement the borrow checker manually (instead of using Cell directly).

Oh, I want to mean [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html). Since it isn't possible to obtain a mutable reference from a `Cell`.

Reply:
Sorry, my bad. It is indeed possible to implement in `RefCell`. And we shouldn't make the lifetime of the reference in the closure parameter explicit. It is now fixed. Thanks.